spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkSessionCatalog.java

```diff
@@ -134,28 +135,34 @@ public Identifier[] listTables(String[] namespace) throws NoSuchNamespaceExcepti
 
   @Override
   public Table loadTable(Identifier ident) throws NoSuchTableException {
-    try {
-      return icebergCatalog.loadTable(ident);
-    } catch (NoSuchTableException e) {
-      return getSessionCatalog().loadTable(ident);
-    }
+    return loadTableInternal(ident, null, null);
   }
 
   @Override
   public Table loadTable(Identifier ident, String version) throws NoSuchTableException {
-    try {
-      return icebergCatalog.loadTable(ident, version);
-    } catch (org.apache.iceberg.exceptions.NoSuchTableException e) {
-      return getSessionCatalog().loadTable(ident, version);
-    }
+    return loadTableInternal(ident, version, null);
   }
 
   @Override
   public Table loadTable(Identifier ident, long timestamp) throws NoSuchTableException {
+    return loadTableInternal(ident, null, timestamp);
+  }
+
+  private Table loadTableInternal(Identifier ident, String version, Long timestamp)
+      throws NoSuchTableException {
     try {
-      return icebergCatalog.loadTable(ident, timestamp);
+      return loadTable(icebergCatalog, ident, version, timestamp);
     } catch (org.apache.iceberg.exceptions.NoSuchTableException e) {
-      return getSessionCatalog().loadTable(ident, timestamp);
+      throw e;
+    } catch (org.apache.iceberg.exceptions.NotFoundException e) {
+      if (loadFromSessionCatalogOnLocationNotFoundEnabled()) {
```

Comment:
Hi @RussellSpitzer , could you please confirm this is how you were expecting? If not, please advise.

Reply:
Sure @RussellSpitzer . 
1. Spark Catalog - Do we need to fix this there? --> Only SparkSessionCatalog has delegate catalog. SparkCatalog only works with iceberg table. In this case, it is not a valid iceberg table since metadata folder does not exist.
2.  Drop table even if metadata cannot be loaded. Maybe a Spark Conf? --> Having spark config is good idea. By default, we can keep this config as false.
3. We can maybe do this by checking for MetadataLocation? --> This is also safe thing to do. I thought of checking this only if spark conf is turned on. 

With my proposed change.. it will not break existing functionality and will not be enabled by default. If spark conf set to true, and table contains metadata_location property ( valid iceberg table earlier) and then only load from metastore and delete it. Once metadata does not exist.. this table cant behave as Iceberg table anymore and safer to delete if user requested

```
  private Table loadTableInternal(Identifier ident, String version, Long timestamp)
      throws NoSuchTableException {
    try {
      return loadTable(icebergCatalog, ident, version, timestamp);
    } catch (org.apache.iceberg.exceptions.NoSuchTableException e) {
      throw e;
    } catch (org.apache.iceberg.exceptions.NotFoundException e) {
      **if (loadFromSessionCatalogOnLocationNotFoundEnabled()) {** // check config is set
        Table table = loadTable(getSessionCatalog(), ident, version, timestamp);
        if (table.properties() != null
            && **table.properties().containsKey(TableProperties.METADATA_LOCATION)**) { // check iceberg table
          return table;
        }
      }
      throw e;
    }
  }

  private boolean loadFromSessionCatalogOnLocationNotFoundEnabled() {
    return Boolean.parseBoolean(
        SparkSession.active()
            .conf()
            .get(
                SparkSQLProperties.LOAD_FROM_SESSION_CATALOG_ON_LOCATION_NOT_FOUND_ENABLED,
                SparkSQLProperties
                    .LOAD_FROM_SESSION_CATALOG_ON_LOCATION_NOT_FOUND_ENABLED_DEFAULT));
  }


```

Reply:
If we have a "notFoundException" doesn't that mean we had to have loaded the MetadataLocation? Doesn't that come from the IO?

The property should probably be named directly to it's purpose. The purpose here is something like "ENABLE_DELETE_WHEN_METADATA_NOT_FOUND"

Reply:
I think this goes the the broader point that we don't want the responsibility here on the "loadTableInternal". This should be the responsibility of the delete method correct?

Reply:
hi @RussellSpitzer drop SQL first calls loadTable method internally . Here is the trace:
```
at org.apache.iceberg.BaseMetastoreTableOperations.refreshFromMetadataLocation(BaseMetastoreTableOperations.java:208)
	at org.apache.iceberg.BaseMetastoreTableOperations.refreshFromMetadataLocation(BaseMetastoreTableOperations.java:185)
	at org.apache.iceberg.BaseMetastoreTableOperations.refreshFromMetadataLocation(BaseMetastoreTableOperations.java:180)
	at org.apache.iceberg.hive.HiveTableOperations.doRefresh(HiveTableOperations.java:176)
	at org.apache.iceberg.BaseMetastoreTableOperations.refresh(BaseMetastoreTableOperations.java:97)
	at org.apache.iceberg.BaseMetastoreTableOperations.current(BaseMetastoreTableOperations.java:80)
	at org.apache.iceberg.BaseMetastoreCatalog.loadTable(BaseMetastoreCatalog.java:47)
	at org.apache.iceberg.spark.SparkCatalog.load(SparkCatalog.java:625)
	at org.apache.iceberg.spark.SparkCatalog.loadTable(SparkCatalog.java:151)
	at org.apache.iceberg.spark.SparkSessionCatalog.loadTable(SparkSessionCatalog.java:415)
	at org.apache.iceberg.spark.SparkSessionCatalog.loadTableInternal(SparkSessionCatalog.java:154)
	at org.apache.iceberg.spark.SparkSessionCatalog.loadTable(SparkSessionCatalog.java:138)
	at org.apache.spark.sql.connector.catalog.CatalogV2Util$.loadTable(CatalogV2Util.scala:311)
	at org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveRelations$.$anonfun$lookupTableOrView$2(Analyzer.scala:1134)
	at scala.Option.orElse(Option.scala:447)
```

Ref: https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DropTableExec.scala

Reply:
Another option is, overriding DropTableExec Physical plan from IcebergSparkSessionExtensions and handle FileNotFound exception